stdlib/asm/mem.masm

```diff
@@ -0,0 +1,44 @@
+# ===== MEMORY FUNCTIONS ==========================================================================
+
+#! Copies `n` words from `read_ptr` to `write_ptr`.
+#!
+#! Stack transition looks as follows:
+#! [n, read_ptr, write_ptr, ...] -> [...]
+#! cycles: 12 + 16n
+export.memcopy
+  # The loop variable is changed with an add instead of sub because the former
+  # uses one fewer cycle. So here the counter is negated. (1 cycles)
+  # stack: [-n, read_ptr, write_ptr, ...]
+  neg
+
+  # Pad the stack because mem_load overwrites it (4 cycles)
+  # stack: [PAD, -n, read_ptr, write_ptr, ...]
+  padw
+
+  # check loop condition (3 cycles)
+  # stack: [b, PAD, -n, read_ptr, write_ptr, ...]
+  dup.4 neq.0
+
+  # LOOP: [PAD, -n, read_ptr, write_ptr, ...]
+  # while(n>0) (16 cycles)
+  while.true
+    # perform read (2 cycles)
+    # stack: [A, -n, read_ptr, write_ptr, ...]
+    dup.5 mem_loadw
+
+    # peform write (2 cycles)
+    # stack: [A, -n, read_ptr, write_ptr, ...]
+    dup.6 mem_storew
+
+    # update counters (9 cycles)
+    swapw # stack: [-n, read_ptr+1, write_ptr+1, x, PAD, ...]
+        add.1 movup.3 movup.3 add.1 movup.3 add.1 movup.3
+    swapw # stack: [PAD, -n, read_ptr+1, write_ptr+1, x, ...]
+
+    dup.4 neq.0 # while(n>0) (3 cycles)
+  end
+
+  # clean stack (4 cycles)
+  # stack: [...]
+  dropw dropw
```

Comment:
`dropw` takes 4 cycles - so, this should probably be 8 cycles.

Also, I think we are dropping 1 extra element from the stack (the `x`). Is this intentional?

Reply:
Oh, nice catch! I did commit some of the changes from testing our previous comments. Thanks!